fix(curriculum): Improve verbiage and add examples to the map, filter, and reduce challenges (#37709)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/use-the-reduce-method-to-analyze-data.english.md

## Description
<section id='description'>

<code>Array.prototype.reduce()</code>, or simply <code>reduce()</code>, is the most general of all array operations in JavaScript. You can solve almost any array processing problem using the <code>reduce</code> method.

> The <code>reduce</code> method allows for more general forms of array processing, and it's possible to show that both <code>filter</code> and <code>map</code> can be derived as a special application of <code>reduce</code>.

Reviewer: ```suggestion
The <code>reduce</code> method allows for more general forms of array processing, and it's possible to show that both <code>filter</code> and <code>map</code> can be derived as special applications of <code>reduce</code>.
```

> The <code>reduce</code> method iterates over each item in an array and returns a single value (i.e. string, number, object, array).  This is acheived via a callback function that returns a value created by the user's logic after each iteration.

Reviewer: ```suggestion
The <code>reduce</code> method iterates over each item in an array and returns a single value (i.e. string, number, object, array).  This is achieved via a callback function that is called each iteration.
```

> The <code>reduce</code> method iterates over each item in an array and returns a single value (i.e. string, number, object, array).  This is achieved via a callback function that is called each iteration.

Reviewer 2: ```suggestion
The <code>reduce</code> method iterates over each item in an array and returns a single value (i.e. string, number, object, array).  This is achieved via a callback function that is called on each iteration.
```

Author: @Manish-Giri Should it be `on each` or `in each`?

Reviewer 2: @RandellDawson Either works, up to you. `during each` is also an option.

> The callback function accepts four arguments (with names of your choice).  While iterating over the array, the arguments can be referenced in the callback function.  The first argument is known as the accumulator, which gets assigned the return value of the callback function during each iteration.  The second argument is the current element being processed.  The third argument is the index of the current element being processed.  The fourth argument is the actual array upon which the <code>reduce</code> method was called.

Reviewer: ```suggestion
The callback function accepts four arguments.  The first argument is known as the accumulator, which gets assigned the return value of the callback function from the previous iteration.  The second is the current element being processed.  The third is the index of that element.  The fourth is the array upon which <code>reduce</code> was called.
```

> The callback function accepts four arguments.  The first argument is known as the accumulator, which gets assigned the return value of the callback function from the previous iteration.  The second is the current element being processed.  The third is the index of that element.  The fourth is the array upon which <code>reduce</code> was called.

Reviewer 3: ```suggestion
The callback function accepts four arguments: The first argument is known as the accumulator, which gets assigned the return value of the callback function from the previous iteration, the second is the current element (the element being processed), the third is the index of that element and the fourth is the array upon which <code>reduce</code> is called.
```

Author: I am not sure about replace the period with a colon here.  Not sure why we can not just leave it as a period.  It is a simple sentence as it is.

Reviewer 3: Looks better now?

Author: I honestly don't see a problem with what was originally there.

Reviewer 3: I felt like the statement you had chosen to describe the second parameter is a bit confusing. See the change I’ve made to that part of the sentence.

Author: I feel using `current element` and then in parentheses using the word `the element` is redundant.  I am OK with the commas and changing the `was` to `is` in the last part of the sentence though.

Reviewer 3: Edited.

> In addition to the callback function, <code>reduce</code> has an additional parameter which takes an initial value for the accumulator.  If this second parameter is not used, then the first iteration is skipped and the second iteration gets passed the first element of the array as the accumulator.

Reviewer 3: These lines have two spaces after each sentence. Please fix that.

Author: Not sure what you mean.  See below for how it renders locally.

![image](https://user-images.githubusercontent.com/5313213/68605444-4558ae80-0461-11ea-9587-15161a831b58.png)

> In addition to the callback function, the <code>reduce</code>method has an additional parameter which takes an initial value for the accumulator.  If this second parameter is not used, then the accumulator's initial value is the first element of the array and the second argument is the second element in the first iteration.

Reviewer: ```suggestion
In addition to the callback function, <code>reduce</code> has an additional parameter which takes an initial value for the accumulator.  If this second parameter is not used, then the first iteration is skipped and the second iteration gets passed the first element of the array as the accumulator.
```

> See below for an example using the <code>reduce</code> method on a <code>users</code> array containing objects to return the sum of all the users' ages. The example only uses the first and second arguments, but there are times when you may need to reference the third and/or fourth arguments.

Reviewer: ```suggestion
See below for an example using <code>reduce</code> on the <code>users</code> array to return the sum of all the users' ages. For simplicity the example only uses the first and second arguments.
```

Author: Can you add a comma after `For simplicity` so I can commit this?

```js
const users = [
  { name: 'John', age: 34 },
  { name: 'Amy', age: 20 },
  { name: 'camperCat', age: 10 }
];

> const u = users.reduce((sum, user) => sum + user.age, 0);

Reviewer 2: ```suggestion
const sumOfAges = users.reduce((sum, user) => sum + user.age, 0);
```
